common/src/main/java/com/proximyst/ban/message/PlaceholderMessage.java

```diff
@@ -0,0 +1,104 @@
+
+public final class PlaceholderMessage implements IMessage {
+  private final @NonNull MessageKey messageKey;
+  private final @NonNull MessagesConfig messagesConfig;
+  private final @NonNull IMessageComponent @NonNull [] messageComponents;
+  private final @NonNull BanExceptionalFutureLogger<?> banExceptionalFutureLogger;
+
+  @AssistedInject
+  public PlaceholderMessage(final @Assisted @NonNull MessageKey messageKey,
+      final @NonNull MessagesConfig messagesConfig,
+      final @Assisted @NonNull IMessageComponent @NonNull [] messageComponents,
+      final @NonNull IBanExceptionalFutureLoggerFactory banExceptionalFutureLoggerFactory) {
+    this.messageKey = messageKey;
+    this.messagesConfig = messagesConfig;
+    this.messageComponents = messageComponents;
+    this.banExceptionalFutureLogger = banExceptionalFutureLoggerFactory.createLogger(this.getClass());
+  }
+
+  @Override
+  public @NonNull CompletableFuture<@Nullable Void> send(final @NonNull Audience audience,
+      final @NonNull Identity source) {
+    return this.component()
+        .thenAccept(component -> audience.sendMessage(source, component))
+        .exceptionally(this.banExceptionalFutureLogger.cast());
+  }
+
+  @Override
+  public @NonNull CompletableFuture<@NonNull Component> component() {
+    return this.awaitPlaceholders(this.messageComponents)
+        .thenApply(placeholders -> MiniMessage.get().parse(this.messageKey.map(this.messagesConfig), placeholders))
+        .exceptionally(this.banExceptionalFutureLogger.cast());
+  }
+
+  private @NonNull CompletableFuture<@NonNull Map<@NonNull String, @NonNull String>> awaitPlaceholders(
+      final @NonNull IMessageComponent @NonNull [] messageComponents) {
+    if (messageComponents.length == 0) {
+      return CompletableFuture.completedFuture(new HashMap<>());
+    }
+
+    final CompletableFuture<?>[] futures = Arrays.stream(messageComponents)
+        .map(IMessageComponent::await)
+        .toArray(CompletableFuture[]::new);
+
+    return CompletableFuture.allOf(futures)
+        .thenCompose($ -> {
+          // What if we got an IMessageComponent([]) as value from `await`?
+          // Well, we need to recursively merge those into our new map!
+          return this.awaitPlaceholders(Arrays.stream(futures)
+              .map(CompletableFuture::join)
+              .filter(IMessageComponent[].class::isInstance)
+              .flatMap(value -> Arrays.stream((IMessageComponent[]) value))
+              .toArray(IMessageComponent[]::new));
+        })
+        .thenApply(map -> {
+          for (int i = 0; i < messageComponents.length; ++i) {
+            final IMessageComponent component = messageComponents[i];
+            final Object value = futures[i].join();
+            if (value instanceof IMessageComponent[]) {
+              continue;
+            }
+
+            map.put(component.name(), String.valueOf(value));
+          }
+
+          return map;
+        });
```

Comment:
thenCompose and thenApply lambdas can be extracted to their own methods, this will make the code easier to understand without having to read all of it.